File: slidestation/learning.py
"""Learns your colour corrections and applies them to new slides.

Every slide you approve is stored as one example: a small set of image features describing the
scan (how faded, how dark, what cast) together with the settings you settled on. A new slide is
matched against those examples with distance-weighted k-nearest-neighbours, so a tray of faded
blue slides gets the treatment you gave the last faded blue ones, while a neutral tray does not.

No training step, no dependencies: it is a few hundred numbers on disk and runs in microseconds.
"""
from __future__ import annotations

import json
import os
import time
from pathlib import Path

import cv2
import numpy as np

from .imaging import Params
from .store import _atomic_write, library, lock

FEATURES = 14
MIN_EXAMPLES = 5  # below this, stick to the defaults
K = 7
MAX_EXAMPLES = 20_000
# how far a neighbour may be (in standardised feature space) before it stops counting
MAX_DISTANCE = 3.0
LEARNED_KEYS = ("strength", "brightness", "contrast", "warmth", "tint", "saturation")


def features(rgb: np.ndarray, scans: int = 1) -> list[float]:
    """Describe a blended slide in a way that captures fading, cast and contrast."""
    a = cv2.resize(rgb, (160, 107), interpolation=cv2.INTER_AREA).astype(np.float32)
    flat = a.reshape(-1, 3)
    p = np.percentile(flat, [1, 50, 99], axis=0)  # per channel low/mid/high
    lum = flat.mean(1)
    lo, mid, hi = np.percentile(lum, [5, 50, 95])
    eps = 1e-3
    med = p[1]
    return [
        *p.reshape(-1).tolist(),  # 9: per-channel 1/50/99 percentiles
        float(mid),  # overall brightness
        float(hi - lo),  # contrast
        float(np.log((med[0] + eps) / (med[1] + eps))),  # red/green cast
        float(np.log((med[2] + eps) / (med[1] + eps))),  # blue/green cast
        float(min(scans, 5) / 5),  # single scan vs a deep bracket stack
    ]


class Model:
    """Examples + standardisation, persisted as one small JSON file in the library."""

    def __init__(self, path: Path | None = None):
        self.path = path or (library() / "learning.json")
        self.examples: list[dict] = []
        self._X: np.ndarray | None = None
        self._mu = self._sd = None
        self.load()

    # ------------------------------------------------------------------ storage
    def load(self) -> None:
        try:
            data = json.loads(self.path.read_text())
            self.examples = data.get("examples", [])
        except (OSError, ValueError):
            self.examples = []
        self._fit()

    def save(self) -> None:
        with lock:
            _atomic_write(self.path, {"version": 1, "examples": self.examples[-MAX_EXAMPLES:]})

    def _fit(self) -> None:
        if len(self.examples) < MIN_EXAMPLES:
            self._X = None
            return
        X = np.array([e["f"] for e in self.examples], dtype=np.float32)
        self._mu = X.mean(0)
        self._sd = X.std(0) + 1e-3
        self._X = (X - self._mu) / self._sd

    # ------------------------------------------------------------------ use
    def remember(self, key: str, feats: list[float], params: dict) -> None:
        """Record (or update) the settings a slide was approved with."""
        if len(feats) != FEATURES:
            return
        entry = {"key": key, "f": [round(float(x), 5) for x in feats],
                 "p": {k: float(params[k]) for k in LEARNED_KEYS if k in params},
                 "trim": bool(params.get("trim", True)), "t": time.time()}
        with lock:
            for i, e in enumerate(self.examples):
                if e.get("key") == key:
                    self.examples[i] = entry
                    break
            else:
                self.examples.append(entry)
        self.save()
        self._fit()

    def forget(self, key: str) -> None:
        with lock:
            n = len(self.examples)
            self.examples = [e for e in self.examples if e.get("key") != key]
        if len(self.examples) != n:
            self.save()
            self._fit()

    def suggest(self, feats: list[float]) -> tuple[dict | None, int]:
        """Predict settings for a new slide. Returns (params or None, neighbours used)."""
        if self._X is None or len(feats) != FEATURES:
            return None, 0
        q = (np.array(feats, dtype=np.float32) - self._mu) / self._sd
        d = np.sqrt(((self._X - q) ** 2).mean(1))
        idx = np.argsort(d)[: min(K, len(d))]
        idx = [i for i in idx if d[i] <= MAX_DISTANCE]
        if not idx:
            return None, 0
        w = 1.0 / (d[idx] + 0.25)
        w = w / w.sum()
        out: dict = {}
        for k in LEARNED_KEYS:
            vals = np.array([self.examples[i]["p"].get(k, getattr(Params(), k)) for i in idx], dtype=np.float32)
            out[k] = float(np.round((vals * w).sum(), 3))
        trims = np.array([self.examples[i].get("trim", True) for i in idx], dtype=np.float32)
        out["trim"] = bool((trims * w).sum() >= 0.5)
        return out, len(idx)

    def stats(self) -> dict:
        return {
            "examples": len(self.examples),
            "ready": self._X is not None,
            "min_examples": MIN_EXAMPLES,
            "last": max((e.get("t", 0) for e in self.examples), default=0),
        }


_model: Model | None = None


def model() -> Model:
    global _model
    if _model is None or _model.path != (library() / "learning.json"):
        _model = Model()
    return _model


def reset() -> None:
    """Forget everything (used by the Settings screen)."""
    m = model()
    with lock:
        m.examples = []
    m.save()
    m._fit()

File: frontend/src/standalone/learning.ts
// Learns your colour corrections and applies them to new slides (slidestation/learning.py):
// 14 image features per developed slide plus the settings you accepted, distance-weighted k-NN.
// The examples are stored in the library's learning.json in the same format as the Python app's.
import type { Params } from "@/lib/api";
import { percentile, resized, type RGB } from "./pixels";

export const FEATURES = 14;
export const MIN_EXAMPLES = 5; // below this, stick to the defaults
const K = 7;
const MAX_EXAMPLES = 20_000;
const MAX_DISTANCE = 3.0; // in standardised feature space
export const LEARNED_KEYS = ["strength", "brightness", "contrast", "warmth", "tint", "saturation"] as const;
const DEFAULTS: Record<string, number> = { strength: 0.6, brightness: 0, contrast: 0, warmth: 0, tint: 0, saturation: 0 };

/** Describe a blended slide in a way that captures fading, cast and contrast. */
export function features(a: RGB, scans = 1): number[] {
  const s = resized(a, 160, 107);
  const n = s.width * s.height;
  const ch = [0, 1, 2].map((c) => Float32Array.from({ length: n }, (_, i) => s.data[i * 3 + c]).sort());
  const lum = Float32Array.from({ length: n }, (_, i) => (s.data[i * 3] + s.data[i * 3 + 1] + s.data[i * 3 + 2]) / 3).sort();
  const p = ch.map((v) => [1, 50, 99].map((q) => percentile(v, q)));
  const [lo, mid, hi] = [5, 50, 95].map((q) => percentile(lum, q));
  const eps = 1e-3;
  const med = p.map((x) => x[1]);
  return [
    ...[0, 1, 2].flatMap((q) => p.map((c) => c[q])), // 9: per-channel 1/50/99 percentiles, quantile-major like numpy
    mid, // overall brightness
    hi - lo, // contrast
    Math.log((med[0] + eps) / (med[1] + eps)), // red/green cast
    Math.log((med[2] + eps) / (med[1] + eps)), // blue/green cast
    Math.min(scans, 5) / 5, // single scan vs a deep bracket stack
  ];
}

export type Example = { key: string; f: number[]; p: Record<string, number>; trim: boolean; t: number };

/** Examples + standardisation. `save` is called with the JSON to persist after every change. */
export class Model {
  examples: Example[] = [];
  private X: Float32Array[] | null = null;
  private mu: number[] = [];
  private sd: number[] = [];

  constructor(data: unknown, private save: (json: unknown) => void) {
    const ex = (data as { examples?: Example[] } | null)?.examples;
    this.examples = Array.isArray(ex) ? ex : [];
    this.fit();
  }

  private fit() {
    if (this.examples.length < MIN_EXAMPLES) {
      this.X = null;
      return;
    }
    const n = this.examples.length;
    this.mu = new Array(FEATURES).fill(0);
    this.sd = new Array(FEATURES).fill(0);
    for (const e of this.examples) e.f.forEach((v, j) => (this.mu[j] += v / n));
    for (const e of this.examples) e.f.forEach((v, j) => (this.sd[j] += (v - this.mu[j]) ** 2 / n));
    this.sd = this.sd.map((v) => Math.sqrt(v) + 1e-3);
    this.X = this.examples.map((e) => Float32Array.from(e.f, (v, j) => (v - this.mu[j]) / this.sd[j]));
  }

  private persist() {
    this.save({ version: 1, examples: this.examples.slice(-MAX_EXAMPLES) });
    this.fit();
  }

  /** Record (or update) the settings a slide was developed with. */
  remember(key: string, feats: number[], params: Params) {
    if (feats.length !== FEATURES) return;
    const p: Record<string, number> = {};
    for (const k of LEARNED_KEYS) if (k in params) p[k] = Number(params[k]);
    const entry: Example = { key, f: feats.map((x) => Math.round(x * 1e5) / 1e5), p, trim: params.trim !== false, t: Date.now() / 1000 };
    const i = this.examples.findIndex((e) => e.key === key);
    if (i >= 0) this.examples[i] = entry;
    else this.examples.push(entry);
    this.persist();
  }

  forget(key: string) {
    const n = this.examples.length;
    this.examples = this.examples.filter((e) => e.key !== key);
    if (this.examples.length !== n) this.persist();
  }

  reset() {
    this.examples = [];
    this.persist();
  }

  /** Predict settings for a new slide: the settings, and how many neighbours they came from. */
  suggest(feats: number[]): [Partial<Params> | null, number] {
    if (!this.X || feats.length !== FEATURES) return [null, 0];
    const q = feats.map((v, j) => (v - this.mu[j]) / this.sd[j]);
    const d = this.X.map((x) => Math.sqrt(x.reduce((s, v, j) => s + (v - q[j]) ** 2, 0) / FEATURES));
    const idx = d
      .map((v, i) => i)
      .sort((a, b) => d[a] - d[b])
      .slice(0, Math.min(K, d.length))
      .filter((i) => d[i] <= MAX_DISTANCE);
    if (!idx.length) return [null, 0];
    const w0 = idx.map((i) => 1 / (d[i] + 0.25));
    const ws = w0.reduce((s, v) => s + v, 0);
    const w = w0.map((v) => v / ws);
    const out: Record<string, unknown> = {};
    for (const k of LEARNED_KEYS) {
      const v = idx.reduce((s, i, n) => s + (this.examples[i].p[k] ?? DEFAULTS[k]) * w[n], 0);
      out[k] = Math.round(v * 1000) / 1000;
    }
    out.trim = idx.reduce((s, i, n) => s + (this.examples[i].trim !== false ? 1 : 0) * w[n], 0) >= 0.5;
    return [out as Partial<Params>, idx.length];
  }

  stats() {
    return {
      examples: this.examples.length,
      ready: this.X !== null,
      min_examples: MIN_EXAMPLES,
      last: Math.max(0, ...this.examples.map((e) => e.t || 0)),
    };
  }
}
